hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -722,50 +727,36 @@ private class Emit(
       case _: ArrayMap | _: ArrayZip | _: ArrayFilter | _: ArrayRange | _: ArrayFlatMap | _: ArrayScan | _: ArrayLeftJoinDistinct | _: RunAggScan | _: ReadPartition | _: MakeStream | _: StreamRange =>
         emitArrayIterator(ir).toEmitTriplet(mb, PArray(coerce[PStreamable](ir.pType).elementType))
 
-      case ArrayFold(a, zero, name1, name2, body) =>
-        val typ = ir.typ
-        val tarray = coerce[TStreamable](a.typ)
-        val tti = typeToTypeInfo(typ)
-        val eti = typeToTypeInfo(tarray.elementType)
-        val xmv = mb.newField[Boolean](name2 + "_missing")
-        val xvv = coerce[Any](mb.newField(name2)(eti))
-        val xmbody = mb.newField[Boolean](name1 + "_missing_tmp")
-        val xmaccum = mb.newField[Boolean](name1 + "_missing")
-        val xvaccum = coerce[Any](mb.newField(name1)(tti))
-        val bodyenv = env.bind(
-          (name1, (tti, xmaccum.load(), xvaccum.load())),
-          (name2, (eti, xmv.load(), xvv.load())))
-
-        val codeZ = emit(zero)
-        val codeB = emit(body, env = bodyenv)
-        val aBase = emitArrayIterator(a)
+      case ArrayFold(a, zero, accumName, valueName, body) =>
+        val eltType = a.pType.asInstanceOf[PStreamable].elementType
+        val accType = ir.pType
+        implicit val eltPack = TypedTriplet.pack(eltType)
+        implicit val accPack = TypedTriplet.pack(accType)
+        val accTI = typeToTypeInfo(accType)
+        val eltTI = typeToTypeInfo(eltType)
+
+        val streamOpt = emitStream2(a)
+        val resOpt: COption[Code[_]] = streamOpt.flatMapCPS { (stream, _ctx, ret) =>
+          implicit val c = _ctx
+          def foldBody(elt: TypedTriplet[eltType.type], acc: TypedTriplet[accType.type]): TypedTriplet[accType.type] = {
```

Comment:
TypedTriplet is just a EmitTriplet with a type parameter for static type checking, right?  I have two thoughts about this:

I don't think we should fork a feature like this and have some code use it and some code not use it unless we're planning to convert all the code to the new style/feature.  I have found on large projects aiming for global consistency of the code is very valuable, although of course things are changing and you won't always think it is the best use of time to push the changes through the rest of the code immediately.  Are you in effect proposing to add a type parameter to EmitTriplet?  I will remark, which the PValue change, the EmitTriplet now carries its physical type, which improve type checking for debugging, although not statically.

Reply:
TypedTriplet also came out of Milo's stream work. I don't remember the original motivations, and I agree it feels like technical debt that should be cleaned up somehow. I think the primary reason for carrying the PType is to support making a `ParameterPack[TypedTriplet]`, not necessarily for type checking.

I am taking advantage of this to simplify some of the emit code by replacing to repeated pattern
```
val xm = mb.newField[Boolean]()
val xv = mb.newField()(typeToTypeInfo(typ))
Code(
  et.setup,
  xm := et.m,
  xv := xm.mux(defaultValue(typ), et.v),
  ...)
```
with
```
val x = pp.newFields(mb.fb)
val tt = TypedTriplet(typ, et)
Code(
  x := tt,
  ...)
```
which I find makes the emit code much easier to read. But I think PValues will allow something similar, right?

Reply:
I agree the old code style isn't clean.  PValues by themselves don't fix this, they replace EmitTriplet.v: they represent non-missing values.  They can be stored in locals/fields, but don't carry the missing bit.

 > I think the primary reason for carrying the PType is to support making a ParameterPack[TypedTriplet], not necessarily for type checking.

Ah, great, I agree with this 100%.  I'll pull the load/store interface into the mainline code (but drop the type parameter), remove TypedTriplet.  Then the rest of the Emit code can benefit.  Thanks for clarifying!